agent/engine/containermetadata/types.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2014-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package containermetadata
+
+import (
+	"time"
+
+	"github.com/aws/amazon-ecs-agent/agent/api"
+
+	docker "github.com/fsouza/go-dockerclient"
+)
+
+// dockerDummyClient is a wrapper for the docker interface functions we need
+// We use this as a dummy type to be able to pass in engine.DockerClient to
+// our functions without creating import cycles
+type dockerDummyClient interface {
+	InspectContainer(string, time.Duration) (*docker.Container, error)
+	Version() (string, error)
+}
+
+// NetworkMetadata keeps track of the data we parse from the Network Settings
+// in docker containers
+type NetworkMetadata struct {
+	ports       []api.PortBinding
+	networkMode string
+	gateway     string
+	iPAddress   string
+	iPv6Gateway string
+}
+
+// DockerContainerMD keeps track of all metadata acquired from Docker inspection
+// Has redundancies with engine.DockerContainerMetadata but this packages all
+// docker metadata we want in the service so we can change features easily
+type DockerContainerMD struct {
+	status        string
+	containerID   string
+	containerName string
+	imageID       string
+	imageName     string
+	networkInfo   NetworkMetadata
+}
+
+// TaskMetadata keeps track of all metadata associated with a task
+// provided by AWS, does not depend on the creation of the container
+type TaskMetadata struct {
+	version           string
+	clusterArn        string
+	containerInstance string
+	taskArn           string
+}
+
+// Metadata packages all acquired metadata and is used to format it
+// into JSON to write to the metadata file
+type Metadata struct {
+	Version           string            `json:"DockerVersion, omitempty"`
```

Comment:
Do you mean the variable or the tag? If you mean the tag, I do not quite understand why this is necessary since I think most of the code base has capitalized JSON tags. If you mean the variable names, MarshalJSON does not encode unexported variables so either I implement MarshalJSON for Metadata or have these capitalized as I have now. I do not think it is worthwhile to implement MarshalJSON over directly encoding the Metadata struct because I am not doing anything special in the marshaling.